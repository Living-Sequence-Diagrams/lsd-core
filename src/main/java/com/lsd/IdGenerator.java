package com.lsd;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

import static java.util.UUID.randomUUID;

public class IdGenerator {
    private final AtomicInteger counter = new AtomicInteger();

    private final Supplier<String> idSupplier;

    public IdGenerator(boolean isDeterministic) {
        if (isDeterministic) {
            idSupplier = () -> String.valueOf(counter.incrementAndGet());
        } else {
            idSupplier = () -> randomUUID().toString()
                    .replaceAll("-", "");
        }
    }

    public String next() {
        return idSupplier.get();
    }

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    public void reset() {
        counter.set(0);
    }
}
