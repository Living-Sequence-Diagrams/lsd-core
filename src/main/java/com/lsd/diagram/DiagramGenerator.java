package com.lsd.diagram;

import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Participant;
import com.lsd.report.model.SequenceDiagram;
import com.lsd.events.SequenceEvent;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<SequenceEvent> events;

    public SequenceDiagram sequenceDiagram() {
        String uml = generateSequenceUml();
        String svg = generateSequenceSvg(uml);
        return SequenceDiagram.builder()
                .uml(uml)
                .svg(svg)
                .build();
    }

    private String generateSequenceSvg(String markup) {
        SvgConverter svgConverter = new SvgConverter();

        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(SequenceEvent::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
