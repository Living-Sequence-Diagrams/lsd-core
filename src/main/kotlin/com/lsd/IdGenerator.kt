package com.lsd

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Supplier

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    
    private val idSupplier: Supplier<String> = if (isDeterministic)
        Supplier { counter.incrementAndGet().toString() }
    else
        Supplier { randomUUID().toString().replace("-".toRegex(), "") }


    operator fun next(): String = idSupplier.get()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}