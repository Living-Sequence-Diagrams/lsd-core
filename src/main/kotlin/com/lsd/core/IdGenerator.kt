package com.lsd.core

import java.util.UUID.randomUUID
import java.util.concurrent.atomic.AtomicInteger

private val randomIdSupplier: () -> String = { randomUUID().toString().replace("-".toRegex(), "") }

class IdGenerator(isDeterministic: Boolean = false) {
    private val counter = AtomicInteger()
    private val deterministicIdSupplier: () -> String = { counter.incrementAndGet().toString() }

    private val idSupplier: () -> String = if (isDeterministic) deterministicIdSupplier else randomIdSupplier

    fun next(): String = idSupplier.invoke()

    /**
     * Restarts the counter - only works when in deterministic mode
     */
    fun reset() = counter.set(0)
}