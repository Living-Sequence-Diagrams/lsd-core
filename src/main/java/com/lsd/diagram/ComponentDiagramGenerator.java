package com.lsd.diagram;

import com.lsd.IdGenerator;
import com.lsd.events.Message;
import com.lsd.events.SequenceEvent;
import com.lsd.events.SynchronousResponse;
import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Diagram;
import com.lsd.report.model.Participant;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.function.Predicate.not;
import static java.util.stream.Collectors.toList;

@Builder
public class ComponentDiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/component-uml.peb");
    private final SvgConverter svgConverter = new SvgConverter();

    private final List<Participant> participants;
    private final List<SequenceEvent> events;
    private final IdGenerator idGenerator;

    public Diagram diagram() {
        String uml = generateUml();
        return Diagram.builder()
                .id(idGenerator.next())
                .uml(uml)
                .svg(generateSvg(uml))
                .build();
    }

    private String generateSvg(String markup) {
        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "participants", participants.stream()
                        .map(Participant::getMarkup)
                        .map(markup -> markup.replaceAll("^participant ", "component "))
                        .distinct()
                        .collect(toList()),
                "events", events.stream()
                        .filter(event -> event instanceof Message)
                        .filter(not(SynchronousResponse.class::isInstance))
                        .map(event -> (Message) event)
                        .map(message -> String.format("[%s] -> %s", message.getFrom(), message.getTo()))
                        .distinct()
                        .collect(toList())
        ));
        return writer.toString();
    }
}
