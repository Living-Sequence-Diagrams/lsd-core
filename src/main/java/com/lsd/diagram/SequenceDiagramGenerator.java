package com.lsd.diagram;

import com.lsd.IdGenerator;
import com.lsd.events.SequenceEvent;
import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Diagram;
import com.lsd.report.model.Participant;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;
import static org.apache.commons.collections4.ListUtils.partition;

@Builder
public class SequenceDiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<SequenceEvent> events;
    private final IdGenerator idGenerator;

    public Optional<Diagram> diagram(int maxEventsPerDiagram) {
        if (events.isEmpty()) {
            return Optional.empty();
        }
        var uml = partition(events, maxEventsPerDiagram).stream()
                .map(this::generateSequenceUml)
                .collect(joining(System.lineSeparator()));

        String svg = generateSequenceSvg(uml);
        return Optional.of(Diagram.builder()
                .id(idGenerator.next())
                .uml(uml)
                .svg(svg)
                .build());
    }

    private String generateSequenceSvg(String markup) {
        SvgConverter svgConverter = new SvgConverter();

        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateSequenceUml(List<SequenceEvent> events) {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants.stream().distinct().collect(toList()),
                "events", events.stream()
                        .map(SequenceEvent::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
