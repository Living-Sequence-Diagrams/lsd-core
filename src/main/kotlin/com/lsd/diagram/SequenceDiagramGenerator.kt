package com.lsd.diagram

import com.github.jknack.handlebars.Handlebars
import com.lsd.IdGenerator
import com.lsd.builders.SequenceDiagramGeneratorBuilder
import com.lsd.events.SequenceEvent
import com.lsd.events.groupedByPages
import com.lsd.properties.LsdProperties
import com.lsd.properties.LsdProperties.DIAGRAM_THEME
import com.lsd.report.model.Diagram
import com.lsd.report.model.Participant
import org.apache.commons.collections4.ListUtils
import java.lang.System.lineSeparator
import java.util.stream.Collectors.joining
import java.util.stream.Collectors.toList

data class SequenceDiagramGenerator(
    private val idGenerator: IdGenerator,
    private val events: List<SequenceEvent> = emptyList(),
    private val includes: List<String> = emptyList(),
    private val participants: List<Participant> = emptyList(),
) {
    private val template = Handlebars().compile("templates/sequence-uml")

    fun diagram(maxEventsPerDiagram: Int): Diagram? {
        val uml = generateUml(maxEventsPerDiagram)
        return if (events.isEmpty()) null else {
            Diagram(
                id = idGenerator.next(),
                uml = uml,
                svg = convertToSvg(markup = uml)
            )
        }
    }

    private fun generateUml(maxEventsPerDiagram: Int): String =
        events.groupedByPages()
            .stream()
            .flatMap { ListUtils.partition(it, maxEventsPerDiagram).stream() }
            .map { generateSequenceUml(it) }
            .collect(joining(lineSeparator()))

    private fun generateSequenceUml(events: List<SequenceEvent>): String {
        return template.apply(
            mapOf(
                "theme" to LsdProperties.get(DIAGRAM_THEME),
                "includes" to includes.distinct(),
                "participants" to participants.stream()
                    .distinct()
                    .collect(toList()),
                "events" to events.stream()
                    .map(SequenceEvent::toMarkup)
                    .collect(toList())
            )
        )
    }

    companion object {
        fun builder(): SequenceDiagramGeneratorBuilder = SequenceDiagramGeneratorBuilder()
    }
}

