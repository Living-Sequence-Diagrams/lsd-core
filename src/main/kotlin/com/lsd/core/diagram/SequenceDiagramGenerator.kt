package com.lsd.core.diagram

import com.lsd.core.IdGenerator
import com.lsd.core.adapter.puml.convertToSvgAsync
import com.lsd.core.adapter.puml.toParticipantMarkup
import com.lsd.core.adapter.puml.toPumlMarkup
import com.lsd.core.domain.*
import com.lsd.core.properties.LsdProperties
import com.lsd.core.properties.LsdProperties.DIAGRAM_THEME
import com.lsd.core.report.HandlebarsWrapper
import com.lsd.core.report.model.Diagram
import java.lang.System.lineSeparator


data class SequenceDiagramGenerator(
    private val idGenerator: IdGenerator,
    private val events: List<SequenceEvent> = emptyList(),
    private val includes: List<String> = emptyList(),
    private val participants: List<Participant> = emptyList(),
) {
    private val template = HandlebarsWrapper.compile("templates/puml/sequence-uml")

    fun diagram(maxEventsPerDiagram: Int, isDevMode: Boolean): Diagram? {
        if (events.isEmpty()) return null

        val uml = generateUml(maxEventsPerDiagram, isDevMode)
        val svg = convertToSvgAsync(markup = uml)
        return Diagram(id = idGenerator.next(), uml = uml, svg = svg)
    }

    private fun generateUml(maxEventsPerDiagram: Int, isDevMode: Boolean): String {
        return removeLifelineActivationsIfDiagramWillBeSplit(maxEventsPerDiagram)
            .sortedBy { it.created }
            .groupedByPages()
            .flatMap { it.chunked(maxEventsPerDiagram) }
            .joinToString(separator = lineSeparator()) { generateSequenceUml(it, isDevMode) }
    }

    private fun removeLifelineActivationsIfDiagramWillBeSplit(maxEventsPerDiagram: Int) =
        if (events.size > maxEventsPerDiagram || events.any { it is Newpage })
            events.filterNot {
                it is ActivateLifeline || it is DeactivateLifeline
            }
        else events

    private fun generateSequenceUml(events: List<SequenceEvent>, isDevMode: Boolean): String {
        if (events.none { it is Message }) return ""
        return template.apply(
            mapOf(
                "theme" to LsdProperties[DIAGRAM_THEME],
                "includes" to includes.distinct(),
                "participants" to participants.usedIn(events).map(Participant::toParticipantMarkup),
                "events" to events.map(SequenceEvent::toPumlMarkup),
                "devMode" to isDevMode,
            )
        )
    }
}

fun Collection<Participant>.usedIn(events: List<SequenceEvent>): List<Participant> {
    val suppliedParticipantNames = associate { it.componentName.normalisedName to it }
    return events.filterIsInstance<Message>()
        .flatMap { listOf(it.from, it.to) }
        .distinctBy { it.componentName.normalisedName }
        .associateBy { it.componentName.normalisedName }
        .filter { it.key.isNotBlank() }
        .map { eventParticipant ->
            suppliedParticipantNames.getOrDefault(
                key = eventParticipant.key,
                defaultValue = eventParticipant.value
            )
        }
}

fun List<SequenceEvent>.groupedByPages() =
    fold(mutableListOf<MutableList<SequenceEvent>>()
        .also { it.add(ArrayList()) }) { groups, event ->
        when (event) {
            is Newpage -> groups.apply {
                add(ArrayList())
                last().add(event.pageTitle)
            }

            else -> groups.apply { last().add(event) }
        }
    }

