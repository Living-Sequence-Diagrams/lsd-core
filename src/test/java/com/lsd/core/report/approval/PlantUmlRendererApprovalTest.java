package com.lsd.core.report.approval;

import com.lsd.core.IdGenerator;
import com.lsd.core.diagram.SequenceDiagramGenerator;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Objects;

import static com.lsd.core.builders.MessageBuilder.messageBuilder;
import static com.lsd.core.domain.MessageType.BI_DIRECTIONAL;
import static com.lsd.core.domain.MessageType.SYNCHRONOUS_RESPONSE;
import static com.lsd.core.domain.ParticipantType.*;
import static org.approvaltests.Approvals.verify;

class PlantUmlRendererApprovalTest {

    @Test
    void renderSequenceUml() {
        var diagram = Objects.requireNonNull(sequenceDiagramGenerator().diagram(2));
        verify(diagram.getUml());
    }

    private SequenceDiagramGenerator sequenceDiagramGenerator() {
        return new SequenceDiagramGenerator(
                new IdGenerator(true),
                List.of(
                        messageBuilder()
                                .id("1")
                                .label(" Sending a request ")
                                .from("A")
                                .to("B")
                                .data("Please help?")
                                .build(),
                        messageBuilder()
                                .id("2")
                                .label("Sending a response with a long label that should be abbreviated if all goes well")
                                .from("B")
                                .to("A")
                                .data("OK")
                                .type(SYNCHRONOUS_RESPONSE)
                                .build(),
                        messageBuilder()
                                .id("3")
                                .label("On the phone")
                                .from("B")
                                .to("A")
                                .type(BI_DIRECTIONAL)
                                .build()),
                List.of(
                        "tupadr3/font-awesome-5/clock",
                        "tupadr3/font-awesome-5/database"
                ),
                List.of(
                        ACTOR.called("Arnold", "Arnie"),
                        BOUNDARY.called("Nick"),
                        CONTROL.called("A control"),
                        COLLECTIONS.called("some collection"),
                        ENTITY.called("Entity"),
                        PARTICIPANT.called("Party"),
                        QUEUE.called("Party"),
                        QUEUE.called("Party"), // Duplicate participant to test dropping duplicates
                        DATABASE.called("Derek", "D for danger")
                ));
    }
}