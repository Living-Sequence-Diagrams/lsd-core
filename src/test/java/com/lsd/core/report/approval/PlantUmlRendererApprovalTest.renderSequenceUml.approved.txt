@startuml

!theme lsd-light from https://raw.githubusercontent.com/lsd-consulting/lsd-core/main/src/main/resources/static

hide unlinked


!include <tupadr3/font-awesome-5/clock>

!include <tupadr3/font-awesome-5/database>





A -> B: [[#1 {Sending a request} Sending a request]]

B --> A: [[#2 {Sending a response with a long label that should be abbreviated if all goes well} Sending a response with a long label that shoul...]]


@enduml
@startuml

!theme lsd-light from https://raw.githubusercontent.com/lsd-consulting/lsd-core/main/src/main/resources/static

hide unlinked


!include <tupadr3/font-awesome-5/clock>

!include <tupadr3/font-awesome-5/database>





B <-> A: [[#3 {On the phone} On the phone]]


@enduml